exporters/metric/cortex/config.go

```diff
@@ -0,0 +1,104 @@
+package cortex
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/spf13/viper"
+)
+
+// Config contains properties the Exporter uses to export metrics data to Cortex.
+type Config struct {
+	Endpoint        string            `mapstructure:"url"`
+	RemoteTimeout   string            `mapstructure:"remote_timeout"`
+	Name            string            `mapstructure:"name"`
+	BasicAuth       map[string]string `mapstructure:"basic_auth"`
+	BearerToken     string            `mapstructure:"bearer_token"`
+	BearerTokenFile string            `mapstructure:"bearer_token_file"`
+	TLSConfig       map[string]string `mapstructure:"tls_config"`
+	ProxyURL        string            `mapstructure:"proxy_url"`
+	PushInterval    string            `mapstructure:"push_interval"`
+	Client          *http.Client
+}
+
+// Option sets an option for a Config struct.
+type Option interface {
+	Apply(*Config)
+}
+
+// WithFilepath adds a path where Viper will search for the YAML file in.
+func WithFilepath(filepath string) Option {
+	return filepathOption(filepath)
+}
+
+type filepathOption string
+
+func (o filepathOption) Apply(config *Config) {
+	viper.AddConfigPath(string(o))
+}
+
+// WithClient adds a custom http.Client to the Config struct.
+func WithClient(client *http.Client) Option {
+	return clientOption{client}
+}
+
+type clientOption struct {
+	client *http.Client
+}
+
+func (o clientOption) Apply(config *Config) {
+	config.Client = (*http.Client)(o.client)
+}
+
+// NewConfig creates a Config struct with a YAML file and applies Option functions to the Config
+// struct.
+func NewConfig(filename string, opts ...Option) (*Config, error) {
+	var config Config
+	for _, opt := range opts {
+		opt.Apply(&config)
+	}
+
+	viper.SetConfigName(filename)
+	viper.SetConfigType("yaml")
+	viper.SetConfigName(filename)
+	viper.AddConfigPath(".")
```

Comment:
You've provided a fluent `Option`s configuration mechanism, but it seems this would be overriding a user-provided Option. For example if I call `NewConfig("myFile", WithFilepath("myFilePath"))` we're applying `opts` and then overriding the value. I think this call should happen before applying `opts` to the `Config` struct to ensure defaults?

Reply:
Thanks for the feedback! I don't think the order matters here as Viper can use multiple search paths so both `"."` and `"myFilePath"` will be searched. I definitely see how other Options (if more are added in the future) can be overridden though. I'll move the for loop to below the calls.